analyzers/src/SonarAnalyzer.CSharp/Rules/ReferenceEqualityCheckWhenEqualsExists.cs

```diff
@@ -91,63 +88,35 @@ protected override void Initialize(SonarAnalysisContext context)
 
-        private static bool HasTypeConstraintsWhichMightOverrideEquals(ITypeSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals)
-        {
-            if (type.TypeKind != TypeKind.TypeParameter)
-            {
-                return false;
-            }
+        private static bool HasTypeConstraintsWhichMightOverrideEquals(ITypeSymbol type, ISet<INamedTypeSymbol> allInterfacesWithImplementationsOverriddenEquals) =>
+            type.TypeKind == TypeKind.TypeParameter
+            && type is ITypeParameterSymbol typeParameter
+            && typeParameter.ConstraintTypes.Any(t => MightOverrideEquals(t, allInterfacesWithImplementationsOverriddenEquals));
 
-            var typeParameter = (ITypeParameterSymbol)type;
-            return typeParameter.ConstraintTypes.Any(t => MightOverrideEquals(t, allInterfacesWithImplementationsOverriddenEquals));
-        }
+        private static bool IsAllowedType(ITypeSymbol type) => type.IsAny(AllowedTypes) || HasAllowedBaseType(type);
 
-        private static bool IsAllowedType(ITypeSymbol type)
-        {
-            return type.IsAny(AllowedTypes) || HasAllowedBaseType(type);
-        }
+        private static bool HasAllowedBaseType(ITypeSymbol type) =>
+            type.GetSelfAndBaseTypes().Any(t => t.IsAny(AllowedTypesWithAllDerived));
 
-        private static bool HasAllowedBaseType(ITypeSymbol type)
-        {
-            var currentType = type;
-            while (currentType != null)
-            {
-                if (currentType.IsAny(AllowedTypesWithAllDerived))
-                {
-                    return true;
-                }
-                currentType = currentType.BaseType;
-            }
-            return false;
-        }
-
-        private static bool IsBinaryCandidateForReporting(BinaryExpressionSyntax binary, SemanticModel semanticModel)
-        {
-            var equalitySymbol = semanticModel.GetSymbolInfo(binary).Symbol as IMethodSymbol;
-
-            return equalitySymbol.IsInType(KnownType.System_Object) &&
-                !IsInEqualsOverride(semanticModel.GetEnclosingSymbol(binary.SpanStart) as IMethodSymbol);
-        }
+        private static bool IsBinaryCandidateForReporting(BinaryExpressionSyntax binary, SemanticModel semanticModel) =>
+            semanticModel.GetSymbolInfo(binary).Symbol is IMethodSymbol equalitySymbol
+            && equalitySymbol.IsInType(KnownType.System_Object)
+            && !IsInEqualsOverride(semanticModel.GetEnclosingSymbol(binary.SpanStart) as IMethodSymbol);
 
-        private static bool HasEqualsOverride(ITypeSymbol type)
-        {
-            return GetEqualsOverrides(type).Any(m => m.OverriddenMethod.IsInType(KnownType.System_Object));
-        }
+        private static bool HasEqualsOverride(ITypeSymbol type) =>
+            GetEqualsOverrides(type).Any(m => m.OverriddenMethod.IsInType(KnownType.System_Object));
```

Comment:
_Nitpick:_ Align with current coding style
```suggestion
            GetEqualsOverrides(type).Any(x => x.OverriddenMethod.IsInType(KnownType.System_Object));
```